### PR TITLE
[Android] Fix control flow issues

### PR DESCRIPTION
**Pull request by adamradocz**
**Description of the Change**
_Fix Redundant 'if' statement inspections_
Reports if statements which can be simplified to single assignment, return or assert statements.

_Fix Simplifiable conditional expression inspection_
Reports conditional expressions of the following form
  condition ? true : foo
  condition ? false : foo
  condition ? foo : !foo
  condition ? !foo : foo
These expressions may respectively be safely simplified to
  condition || foo
  !condition && foo
  condition == foo
  condition != foo

**Release Notes**
N/A

### CODE REVIEW

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectDetailsFragment.java [hunk 1]
```diff
@@ -361,10 +361,9 @@ protected Boolean doInBackground(Void... params) {
 				return false;
 			}
 			// load slideshow images
-//			slideshowImages = status.getSlideshowForProject(project.master_url);
-			if(slideshowImages == null || slideshowImages.size() == 0) return false;
-			return true;
-		}
+            // slideshowImages = status.getSlideshowForProject(project.master_url);
+            return slideshowImages != null && slideshowImages.size() != 0;
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectDetailsFragment.java]: When `return`ing the result of a non-trivial logic expression like this, I think you should put the expression into parentheses. It just reduces confusion to those fellow coders who are not yet used to this style of handling conditionals.

I'll refer to this comment in other changes below when I think they could use *parentheses* as well.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientInterfaceImplementation.java [hunk 2]
```diff
@@ -172,7 +172,7 @@ public Boolean attachProject(String url, String projectName, String authenticato
     			try {Thread.sleep(minRetryInterval);} catch (Exception e) {}
     			reply = projectAttachPoll();
     		}
-    		if(reply != null && reply.error_num == BOINCErrors.ERR_OK) return true;
+            return reply != null && reply.error_num == BOINCErrors.ERR_OK;
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientInterfaceImplementation.java]: *parentheses*

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java [hunk 3]
```diff
@@ -118,11 +118,8 @@ private Boolean determineUserActive() throws Exception {
 		
 		if(telStatus != TelephonyManager.CALL_STATE_IDLE) {
 			newUserActive = true;
-		} else if(screenOn && appPrefs.getSuspendWhenScreenOn() && !appPrefs.getStationaryDeviceMode()) {
-			newUserActive = true;
-		} else {
-			newUserActive = false;
-		}
+		} else
+            newUserActive = screenOn && appPrefs.getSuspendWhenScreenOn() && !appPrefs.getStationaryDeviceMode();
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java]: Please try to be consistent with your use of curly braces. The `if` clauses uses them while the `else` clause doesn't (and line 122 isn't correctly indented making it even more confusing).

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java [hunk 4]
```diff
@@ -111,12 +111,10 @@ else if (localName.equalsIgnoreCase("uses_username")) {
 						mProjectConfig.usesName = true;
 					}
 					else if (localName.equalsIgnoreCase("web_stopped")) {
-						mProjectConfig.webStopped = false; //default in case parsing fails
-						if(Integer.parseInt(mCurrentElement.toString()) != 0) mProjectConfig.webStopped = true;
+                        mProjectConfig.webStopped = Integer.parseInt(mCurrentElement.toString()) != 0;
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java]: *parentheses* and indentation

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java [hunk 5]
```diff
@@ -111,12 +111,10 @@ else if (localName.equalsIgnoreCase("uses_username")) {
...
 					}
 					else if (localName.equalsIgnoreCase("sched_stopped")) {
-						mProjectConfig.schedulerStopped = false; //default in case parsing fails
-						if(Integer.parseInt(mCurrentElement.toString()) != 0) mProjectConfig.schedulerStopped = true;
+                        mProjectConfig.schedulerStopped = Integer.parseInt(mCurrentElement.toString()) != 0;
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java]: *parentheses* and indentation

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java [hunk 6]
```diff
@@ -280,7 +280,7 @@ public synchronized boolean authorize(String password) {
 	 * @return true if connected to BOINC core client, false if not connected
 	 */
 	public final boolean isConnected() {
-		return (mSocket != null) ? mSocket.isConnected() : false;
+		return (mSocket != null) && mSocket.isConnected();
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java]: *parentheses*
 
Also, I think you can remove the parentheses around `(mSocket != null)` when you add the outer ones.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java [hunk 7]
```diff
@@ -295,14 +295,10 @@ public synchronized boolean connectionAlive() {
...
-				// End of stream reached and no data were received in reply
-				// We assume that socket is closed on the other side,
-				// most probably client shut down
-				return false;
-			}
-			return true;
-		}
+            // If end of stream reached and no data were received in reply
+            // we assume that socket is closed on the other side, most probably client shut down
+            return result.length() != 0;
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java]: *parentheses*

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/VersionInfo.java [hunk 8]
```diff
@@ -32,8 +32,7 @@ public boolean greater_than(VersionInfo vi) {
 		else if (major < vi.major) return false;
 		else if (minor > vi.minor) return true;
 		else if (minor < vi.minor) return false;
-		else if (release > vi.release) return true;
-		else return false;
+		else return release > vi.release;
```
Comment (brevilo) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/VersionInfo.java]: *parentheses*

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java (hunk 3 above)
Comment (AenBleidd) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java]: I think here we should either combine `telStatus != TelephonyManager.CALL_STATE_IDLE` with other statements or just left this if-else block without change

**Comment by adamradocz** [comment]
> Just curious: the change tips you quote (?) in the description, where do they come from? Did you use some kind of code-beautifier or does Android Studio give those recommendations?
Android Studio gives the recommendations. It has a code inspect tool.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectDetailsFragment.java (hunk 1 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectDetailsFragment.java]: Fixed.

Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/ProjectDetailsFragment.java]: Fixed.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientInterfaceImplementation.java (hunk 2 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientInterfaceImplementation.java]: Fixed.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java (hunk 3 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java]: Fixed.
@AenBleidd What do you imagine?

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java (hunk 4 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java]: Fixed.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java (hunk 5 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfigReplyParser.java]: Fixed.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java (hunk 6 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java]: Fixed.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/VersionInfo.java (hunk 8 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/VersionInfo.java]: Fixed.

**Comment by adamradocz** [comment]
Done.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java (hunk 3 above)
Comment (AenBleidd) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java]: Something like this:

`newUserActive = (telStatus != TelephonyManager.CALL_STATE_IDLE || (screenOn && appPrefs.getSuspendWhenScreenOn() && !appPrefs.getStationaryDeviceMode()));`

Comment (AenBleidd) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java]: But either @adamradocz or @brevilo find this too complicated to understand, than you initial implementation is ok too

Comment (brevilo) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/DeviceStatus.java]: I think the original is easier to comprehend as boolean algebra isn't everybody's strong suit and just I don't see the benefit of squeezing that additional `or` clause in front of the existing clause. IMHO, it just increases complexity.

**Comment by brevilo** [comment]
@AenBleidd feel free to merge when ready.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java (hunk 7 above)
Comment (adamradocz) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java]: Done.

**Pull request merged**